### ISSUE TITLE
`$metadata` endpoint fails with "An asynchronous operation is already in progress."

### ISSUE THREAD

**Issue by vcaraulean** [OData/AspNetCoreOData#1320]
**Assemblies affected**

    <PackageReference Include="Microsoft.AspNetCore.OData" Version="9.0.0" />

**Describe the bug**

Calling `$metadata` endpoint fails with `System.InvalidOperationException An asynchronous operation is already in progress` error. Some data is returned (streamed), but the response is incomplete, the content is not consistent, varies from call to call.

Stack traces are sometimes different. Some specimens:
```
System.InvalidOperationException
An asynchronous operation is already in progress.
   at System.Xml.XmlAsyncCheckWriter.CheckAsync()
   at System.Xml.XmlAsyncCheckWriter.Flush()
   at Microsoft.OData.ODataMetadataOutputContext.Dispose(Boolean disposing)
   at Microsoft.OData.ODataOutputContext.Dispose()
   at Microsoft.OData.ODataMessageWriter.Dispose(Boolean disposing)
   at Microsoft.OData.ODataMessageWriter.Dispose()
   at Microsoft.AspNetCore.OData.Formatter.ODataOutputFormatterHelper.WriteToStreamAsync(Type type, Object value, IEdmModel model, ODataVersion version, Uri baseAddress, MediaTypeHeaderValue contentType, HttpRequest request, IHeaderDictionary requestHeaders, IODataSerializerProvider serializerProvider)
```

Another is slightly different, which is not surprising as there is probably a race condition somewhere...
```
System.InvalidOperationException
An asynchronous operation is already in progress.
   at System.Xml.XmlAsyncCheckWriter.CheckAsync()
   at System.Xml.XmlAsyncCheckWriter.WriteEndElementAsync()
   at Microsoft.OData.Edm.Csdl.Serialization.EdmModelCsdlSchemaXmlWriter.WriteEndElementAsync()
   at Microsoft.OData.Edm.Csdl.Serialization.EdmModelCsdlSerializationVisitor.VisitEdmSchemaAsync(EdmSchema element, IEnumerable`1 mappings)
   at Microsoft.OData.Edm.Csdl.CsdlXmlWriter.WriteSchemasAsync()
   at Microsoft.OData.Edm.Csdl.CsdlXmlWriter.WriteODataCsdlAsync()
   at Microsoft.OData.Edm.Csdl.CsdlXmlWriter.WriteCsdlAsync()
   at Microsoft.OData.Edm.Csdl.CsdlWriter.TryWriteCsdlAsync(IEdmModel model, XmlWriter writer, CsdlTarget target, CsdlXmlWriterSettings writerSettings)
   at Microsoft.OData.ODataMetadataOutputContext.WriteMetadataDocumentImplementationAsync()
   at Microsoft.OData.ODataMetadataOutputContext.WriteMetadataDocumentAsync()
   at Microsoft.OData.ODataMessageWriter.WriteToOutputAsync(ODataPayloadKind payloadKind, Func`2 writeAsyncAction)
   at Microsoft.AspNetCore.OData.Formatter.Serialization.ODataMetadataSerializer.WriteObjectAsync(Object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)
   at Microsoft.AspNetCore.OData.Formatter.ODataOutputFormatterHelper.WriteToStreamAsync(Type type, Object value, IEdmModel model, ODataVersion version, Uri baseAddress, MediaTypeHeaderValue contentType, HttpRequest request, IHeaderDictionary requestHeaders, IODataSerializerProvider serializerProvider)
```

**Reproduce steps**

Happens on our large-ish solution. I was unable to reproduce the issue with a more reduced reproduction solution.

**Data Model**

We have 2 large EDM models: 
 - first has over 150 entities. `$metadata` returns the schema without issues
 - second is ~80. `$metadata` fails with the error above

I tried to try to understand why only second fails, but been unable to do so. Cannot pinpoint the failure to a specific entity or set.

**Expected behavior**

`$metadata` endpoint allows retrieval of OData metadata.

**Additional context**

The failure has another indirection layer - environment:
 - running the solution/project locally (from Rider): `$metadata` succeeds
 - running from integration test (using TestServer): `$metadata` fails
 - running in a deployed environment (corporate k8s cluster): `$metadata` fails

And success/failures are consistent for each environment.

**Comment by vcaraulean** [comment]
@julealgon It took a while... Management denied sharing our domain classes (without scrambling content) so I had to play with some generated models to get same result. 

The repository to reproduce the issue can be found here: https://github.com/vcaraulean/ODataSerializerAsyncError 

Just hit `/odata/$metadata` few times and usually second or third hit is failing.

I tried to look and debug library code, but with all sync/async jumps and layers it's quite hard to follow. Hopefully someone with a better grasp of codebase or with more experience debugging async code will be able to dig out something...

**Comment by vcaraulean** [comment]
I tried as well to reproduce the issue at the boundary of core OData library. My attempts were not successful so far. It's not an indication that the issue is in AspNetCoreOData code, maybe someone else will be able to locate where the issue is coming from.

**Comment by xuzhg** [comment]
If switch to ASP.NET Core OData 8.x version in https://github.com/vcaraulean/ODataSerializerAsyncError, it seems we don't hit the failing.

So, the changes at https://github.com/OData/odata.net/pull/3006 of ODL 8 could be related.

**Comment by habbes** [comment]
The exception appears to protect against concurrent write to the XmlWriter from multiple async tasks, ensuring that the last async write task is complete before starting another one: https://source.dot.net/#System.Private.Xml/System/Xml/Core/XmlWriterAsync.cs,433e2e7260906610

Makes me wonder whether there's case where multiple threads are trying to write to the same XmlWriter concurrently. If the issue only appears 

On a separate note (not sure if it's related), the first stack trace shows that `ODataOutputFormatterHelper.WriteToStreamAsync()` calls `ODataMessageWriter.Dispose()` which could trigger synchronous I/O in async context and lead to thread starvation issues (I think we could fix that by replacing "using" with `await using`

**Comment by habbes** [comment]
At first glance, it sounds like there could be an issue with multiple threads/requests interfering with each other. We have had cases of potential race conditions raised before: https://github.com/OData/AspNetCoreOData/issues/57

**Comment by julealgon** [comment]
@habbes might be worth checking internal OData service DI registrations. There could be a chance they are being shared across threads/requests when they should not or there is a captive dependency somewhere in a singleton.

I'm usually super skeptical about registering things as "singleton" by default because of this type of problem (even if it can be more memory-efficient when used correctly).

**Comment by WanjohiSammy** [comment]
@vcaraulean This has been resolved in [Microsoft.OData.Core 8.1.0](https://www.nuget.org/packages/Microsoft.OData.Core/8.1.0)

**Comment by vcaraulean** [comment]
Thanks @WanjohiSammy !
Is a new release of `Microsoft.AspNetCore.OData` planned as well?
